stock_helper/models/stock_location.py

```diff
@@ -48,14 +48,44 @@ def get_closest_warehouse(self):
 
         With this methods we will really get the closest warehouse of a location
         """
-        self.ensure_one()
-        location_ids = [int(x) for x in self.parent_path.split("/") if x]
         warehouses = (
             self.env["stock.warehouse"]
-            .search([("view_location_id", "in", location_ids)])
+            .search([])
             .sorted(lambda w: w.view_location_id.parent_path, reverse=True)
         )
-        for warehouse in warehouses:
-            if self.parent_path.startswith(warehouse.view_location_id.parent_path):
-                return warehouse
-        return warehouses.browse()
+        res = {}
+        for location in self:
+            wh = False
+            if location.parent_path:
+                for warehouse in warehouses:
+                    if location.parent_path.startswith(
+                        warehouse.view_location_id.parent_path
+                    ):
+                        wh = warehouse
+                        break
+            res[location.id] = wh
+        return res
+
+    def get_closest_warehouse(self):
+        """Returns closest warehouse for current location."""
+        self.ensure_one()
+        location_and_warehouse = self._get_closest_warehouse()
+        warehouse = location_and_warehouse[self.id]
+        return warehouse or self.env["stock.warehouse"]
```

Comment:
```suggestion
        res = defaultdict(self.env["stock.warehouse"])
        for location in self.filtered("parent_path"):
            for warehouse in warehouses:
                if location.parent_path.startswith(
                    warehouse.view_location_id.parent_path
                ):
                    res[location.id] = warehouse
                    break
        return res

    def get_closest_warehouse(self):
        """Returns closest warehouse for current location."""
        self.ensure_one()
        return  self._get_closest_warehouse()[self.id]
```

Also, I kinda feel we can improve `_get_closest_warehouse` and drop the nested loop here.

Reply:
Applied the recommended change, but didn't want to mess about with the function more as I would first need to get more familiar with the context